src/traces/bar/hover.js

```diff
@@ -19,35 +19,38 @@ module.exports = function hoverPoints(pointData, xval, yval, hovermode) {
         trace = cd[0].trace,
         t = cd[0].t,
         xa = pointData.xa,
-        ya = pointData.ya,
-        barDelta = (hovermode === 'closest') ?
-            t.barwidth / 2 :
-            t.bargroupwidth / 2,
-        barPos;
+        ya = pointData.ya;
 
+    var barPos;
     if(hovermode !== 'closest') barPos = function(di) { return di.p; };
     else if(trace.orientation === 'h') barPos = function(di) { return di.y; };
     else barPos = function(di) { return di.x; };
 
+    var barDelta = (hovermode === 'closest') ?
+        function(i) { return (t.barwidth[i] || t.barwidth) / 2; } :
+        function() { return t.bargroupwidth / 2; };
+
     var dx, dy;
     if(trace.orientation === 'h') {
         dx = function(di) {
             // add a gradient so hovering near the end of a
             // bar makes it a little closer match
             return Fx.inbox(di.b - xval, di.x - xval) + (di.x - xval) / (di.x - di.b);
         };
-        dy = function(di) {
+        dy = function(di, i) {
```

Comment:
I suppose this (providing `i` as an extra arg) is the most flexible way to handle the issue for future extensibility, but the way that matches better with what we do in scatter - and should be slightly better perf - would be to include the width of each point in `di`